gax/operation_unittest.cc

```diff
@@ -0,0 +1,234 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "gax/operation.h"
+#include "google/longrunning/operations.pb.h"
+#include "googletest/include/gtest/gtest.h"
+#include "gax/status.h"
+#include <memory>
+#include <string>
+#include <type_traits>
+
+namespace google {
+namespace gax {
+
+class DummyOperationsStub : public gax::OperationsStub {
+ public:
+  ~DummyOperationsStub() = default;
+
+  gax::Status GetOperation(
+      gax::CallContext& context,
+      google::longrunning::GetOperationRequest const& request,
+      google::longrunning::Operation* response) override {
+    auto info = context.Info();
+    auto get_info = OperationsStub::get_operation_info;
+    EXPECT_EQ(std::string(info.rpc_name), std::string(get_info.rpc_name));
+    EXPECT_EQ(info.rpc_type, get_info.rpc_type);
+    EXPECT_EQ(info.idempotency, get_info.idempotency);
+
+    EXPECT_EQ(request.name(), "test");
+    EXPECT_NE(response, nullptr);
+    EXPECT_EQ(response->name(), "test");
+    has_gotten = true;
+    return gax::Status{};
+  }
+  gax::Status DeleteOperation(
+      gax::CallContext& context,
+      google::longrunning::DeleteOperationRequest const& request,
+      google::protobuf::Empty* response) override {
+    auto info = context.Info();
+    auto delete_info = OperationsStub::delete_operation_info;
+    EXPECT_EQ(std::string(info.rpc_name), std::string(delete_info.rpc_name));
+    EXPECT_EQ(info.rpc_type, delete_info.rpc_type);
+    EXPECT_EQ(info.idempotency, delete_info.idempotency);
+
+    EXPECT_EQ(request.name(), "test");
+    EXPECT_NE(response, nullptr);
+    has_deleted = true;
+    return gax::Status{};
+  }
+
+  gax::Status CancelOperation(
+      gax::CallContext& context,
+      google::longrunning::CancelOperationRequest const& request,
+      google::protobuf::Empty* response) override {
+    auto info = context.Info();
+    auto cancel_info = OperationsStub::cancel_operation_info;
+    EXPECT_EQ(std::string(info.rpc_name), std::string(cancel_info.rpc_name));
+    EXPECT_EQ(info.rpc_type, cancel_info.rpc_type);
+    EXPECT_EQ(info.idempotency, cancel_info.idempotency);
+
+    EXPECT_EQ(request.name(), "test");
+    EXPECT_NE(response, nullptr);
+    has_canceled = true;
+    return gax::Status{};
+  }
+
+  static void clear_flags() {
+    has_gotten = false;
+    has_deleted = false;
+    has_canceled = false;
+  }
+
+  static bool has_gotten;
+  static bool has_deleted;
+  static bool has_canceled;
+};
+
+bool DummyOperationsStub::has_gotten;
+bool DummyOperationsStub::has_deleted;
+bool DummyOperationsStub::has_canceled;
+
+static_assert(
```

Comment:
It's probably better to test it in the class definition itself so that it better covers the template parameters. Initially, I just wanted to have all 'tests' in the test file.